File: src/graphs/graph.py
from __future__ import annotations


class Graph:
    def __init__(self, directed: bool = False):
        self.directed = directed
        self._adj: dict[str, list[tuple[str, float]]] = {}

    def add_node(self, node: str) -> None:
        if node not in self._adj:
            self._adj[node] = []

    def add_edge(self, u: str, v: str, weight: float = 1.0) -> None:
        self.add_node(u)
        self.add_node(v)
        self._adj[u].append((v, weight))
        if not self.directed:
            self._adj[v].append((u, weight))

    def neighbors(self, node: str) -> list[tuple[str, float]]:
        return self._adj.get(node, [])

    def nodes(self) -> list[str]:
        return list(self._adj.keys())

    def edges(self) -> list[tuple[str, str, float]]:
        seen: set[tuple[str, str]] = set()
        result: list[tuple[str, str, float]] = []
        for u in self._adj:
            for v, w in self._adj[u]:
                key = (min(u, v), max(u, v)) if not self.directed else (u, v)
                if key not in seen:
                    seen.add(key)
                    result.append((u, v, w))
        return result

    def degree(self, node: str) -> int:
        return len(self._adj.get(node, []))

    def has_node(self, node: str) -> bool:
        return node in self._adj

    def has_edge(self, u: str, v: str) -> bool:
        return any(n == v for n, _ in self._adj.get(u, []))

    def subgraph(self, nodes: set[str]) -> Graph:
        g = Graph(directed=self.directed)
        for n in nodes:
            g.add_node(n)
        for u in nodes:
            for v, w in self._adj.get(u, []):
                if v in nodes and not g.has_edge(u, v):
                    g.add_edge(u, v, w)
        return g

    def order(self) -> int:
        return len(self._adj)

    def size(self) -> int:
        return len(self.edges())

    def density(self) -> float:
        n = self.order()
        if n < 2:
            return 0.0
        e = self.size()
        if self.directed:
            return e / (n * (n - 1))
        return (2 * e) / (n * (n - 1))

File: web/app/lib/graph.ts
import { Adjacencia } from "./types";

export class Graph {
  private adj: Map<string, [string, number][]> = new Map();

  addNode(node: string) {
    if (!this.adj.has(node)) this.adj.set(node, []);
  }

  addEdge(u: string, v: string, w = 1.0) {
    this.addNode(u);
    this.addNode(v);
    this.adj.get(u)!.push([v, w]);
    this.adj.get(v)!.push([u, w]);
  }

  nodes(): string[] {
    return [...this.adj.keys()];
  }

  neighbors(node: string): [string, number][] {
    return this.adj.get(node) ?? [];
  }

  degree(node: string): number {
    return this.neighbors(node).length;
  }

  edges(): [string, string, number][] {
    const seen = new Set<string>();
    const result: [string, string, number][] = [];
    for (const [u, nbrs] of this.adj) {
      for (const [v, w] of nbrs) {
        const key = [u, v].sort().join("-");
        if (!seen.has(key)) {
          seen.add(key);
          result.push([u, v, w]);
        }
      }
    }
    return result;
  }

  order(): number {
    return this.adj.size;
  }

  size(): number {
    return this.edges().length;
  }

  density(): number {
    const n = this.order();
    const m = this.size();
    if (n < 2) return 0;
    return (2 * m) / (n * (n - 1));
  }

  subgraph(nodes: Set<string>): Graph {
    const sub = new Graph();
    for (const n of nodes) sub.addNode(n);
    for (const [u, v, w] of this.edges()) {
      if (nodes.has(u) && nodes.has(v)) sub.addEdge(u, v, w);
    }
    return sub;
  }
}

export function buildGraph(adjacencias: Adjacencia[]): Graph {
  const g = new Graph();
  for (const { origem, destino, peso } of adjacencias) {
    g.addEdge(origem, destino, peso);
  }
  return g;
}
